Add Peak CAN driver

Notgnoshi opened the pull request:
I don't have a way of testing this, so I have no idea how far off it is.

Notgnoshi committed:
Add optional tracing feature for development logging

Commit:
Add Peak CAN driver

builds, driver works, no functional test (no pcan available)

---------

Co-authored-by: Jannes Brands <jannes@jannesbrands.de>
Co-authored-by: Austin Gill <notgnoshi@gmail.com>

JannesBrands commented:
I completed the implementation of PCAN that supports all interfaces. Anyway I dont have any PCAN to test it. 
[https://github.com/JannesBrands/AgIsoStack-rs/commit/273dfd6f51b87dbeb63aad20beb0ea5340a03e04](url)

JannesBrands commented on examples/forward.rs:
```diff
@@ -25,13 +62,29 @@ struct Options {
+        #[cfg(feature = "peak")]
+        CanDriver::Pcan => {
+            let bus = parse_usb_bus(iface).unwrap();
+            let baud = ag_iso_stack::driver::Baudrate::Baud250K;
```
Well, the ISO 11783 is running on 250k by default. So I think it makes sense to leave it with the 250k baudrate there.

Notgnoshi committed:
clippy: Fix unwrap-or-default lint

I suspect this lint was added to clippy since we last merged to main,
and that the CI/CD pipeline uses the latest clippy release.

Notgnoshi committed:
Whitespace cleanup

Notgnoshi commented:
# Fedora PCAN-Basic installation instructions

```sh
sudo dnf install kernel-devel popt-devel
tar -xzvf peak-linux-driver-8.16.0.tar.gz
cd peak-linux-driver-8.16.0/
make KERNEL_LOCATION=/usr/src/kernels/6.5.9-200.fc38.x86_64/ clean
EXTRA_CFLAGS=-Wno-error=incompatible-pointer-types make KERNEL_LOCATION=/usr/src/kernels/6.5.9-200.fc38.x86_64/ NET=NETDEV_SUPPORT all
sudo make KERNEL_LOCATION=/usr/src/kernels/6.5.9-200.fc38.x86_64/ install
```
**Question:** Should this be a netdev or chardev network device? The peak driver documentation says they now default to using chardev.

# `forward` example application instructions

Build the `forward` example like so
```sh
cargo run --example forward --features tracing,socketcan,peak --input-driver socketcan --input-interface can0 --output-driver pcan --output-interface USB0
```
This will read from the `can0` socketcan interface (which you can use `cansend can0 123#44556677` to send fake data on) and write to the `USB0` pcan interface. I think it's _maybe_ possible to create a virtual pcan device?

I don't know enough about PCAN to know if that would be easy to read from to test, so I think I'd use `forward` to read from `USB0` and forward to `can1` and use `candump can1` to read back to the console for verification.

Notgnoshi committed:
Document project feature matrix in the README

Notgnoshi committed:
Enable separate input and output driver types

Notgnoshi committed:
Disable pcan in CI tests

Including pcan support in the tests requires the test environment have
libpcanbasic available, which has to be built from source.
